File: src/main/scala/li/cil/oc/common/tileentity/traits/power/AppliedEnergistics2.scala
package li.cil.oc.common.tileentity.traits.power

import net.minecraft.core.Direction
import net.minecraft.nbt.CompoundTag

/**
 * AE2（Applied Energistics 2）能量集成 —— **未移植的降级占位实现**。
 *
 * ==原实现（1.7.10）==
 * 通过 `@Injectable.Interface`（ASM 注入）让本 trait 实现 `appeng.api.networking.IGridHost`，
 * 用 `AEApi.instance.createGridNode(new AppliedEnergistics2GridBlock(this))` 建立电网节点，
 * 每 `Settings.get.tickFrequency` 刻用 `IEnergyGrid#extractAEPower` 抽能，
 * 并通过 `Power.fromAE` / `Power.toAE` 换算单位。
 *
 * ==为什么降级==
 *  - `li.cil.oc.common.asm.**`（ASM 注入层）在 1.21.1 已整体删除，不再能"注入式"实现第三方接口；
 *  - AE2 的 API（`appeng.*`）未随本工程移植，`AEApi` / `IGridNode` / `IEnergyGrid` /
 *    `DimensionalCoord` / `AECableType` 全部不可用；
 *  - `li.cil.oc.common.EventHandler#scheduleAE2Add` 与 `li.cil.oc.integration.util.Power` 也未移植。
 *
 * ==恢复方式==
 * AE2 移植后，把 `IGridHost` 相关实现放回本 trait（或改成 NeoForge Capability 形式接入），
 * 并补回 `Power` 单位换算与 `EventHandler.scheduleAE2Add` 的等价逻辑。
 * 现有的对外方法名（[[getGridNode]] / [[getCableConnectionType]] / [[securityBreak]]）
 * 已经保留，恢复时只需替换返回类型与方法体。
 */
trait AppliedEnergistics2 extends Common {
  // 注意：Scala 的自类型不会被继承，TileEntity 的每个子 trait 都必须重新声明。
  self: net.minecraft.world.level.block.entity.BlockEntity =>

  // TODO(integration.appeng): Mods.AppliedEnergistics2 集成未移植，恒为「未启用」，
  // 因此下面所有更新/存档钩子都不会做任何事。
  private def useAppliedEnergistics2Power() = false

  // ----------------------------------------------------------------------- //

  override def tick(): Unit = {
    super.tick()
    // TODO(integration.appeng): 原实现每 `Settings.get.tickFrequency` 刻调用 updateEnergy()，
    // 逐面通过 `IEnergyGrid#extractAEPower(demand, Actionable.MODULATE, PowerMultiplier.CONFIG)`
    // 抽能并 `Power.fromAE` / `Power.toAE` 换算。
  }

  override protected def initialize(): Unit = {
    super.initialize()
    // TODO(integration.appeng): 原实现为 `EventHandler.scheduleAE2Add(this)`，
    // 延迟一 tick 后 `getGridNode(UNKNOWN).updateState()` 把节点加入 AE 电网。
  }

  override def dispose(): Unit = {
    super.dispose()
    // TODO(integration.appeng): 原实现在 `invalidate()` / `onChunkUnload()` 里调用 securityBreak()
    // 销毁 AE 电网节点（`IGridNode#destroy()`）。注意 dispose 可能被调用两次，恢复时需保证幂等。
  }

  // ----------------------------------------------------------------------- //

  override protected def readFromNBTForServer(nbt: CompoundTag): Unit = {
    super.readFromNBTForServer(nbt)
    // TODO(integration.appeng): 原实现为 `getGridNode(UNKNOWN).loadFromNBT("oc:ae2power", nbt)`。
  }

  override protected def writeToNBTForServer(nbt: CompoundTag): Unit = {
    super.writeToNBTForServer(nbt)
    // TODO(integration.appeng): 原实现为 `getGridNode(UNKNOWN).saveToNBT("oc:ae2power", nbt)`。
  }

  // ----------------------------------------------------------------------- //

  /**
   * TODO(integration.appeng): 原返回 `appeng.api.networking.IGridNode`（服务端首次调用时创建并缓存）。
   * AE2 API 不可用时返回 `null`（与客户端分支的返回值一致）。
   */
  def getGridNode(side: Direction): AnyRef = null

  /**
   * TODO(integration.appeng): 原返回 `appeng.api.util.AECableType.SMART`。
   */
  def getCableConnectionType(side: Direction): AnyRef = null

  /**
   * TODO(integration.appeng): 原销毁 AE 电网节点（`getGridNode(UNKNOWN).destroy()`）。
   */
  def securityBreak(): Unit = {}
}

File: mod_src/OpenComputers-main-MC1.21.1/src/main/scala/li/cil/oc/integration/appeng/AppliedEnergistics2.java
package li.cil.oc.integration.appeng;

import appeng.api.config.Actionable;
import appeng.api.config.PowerMultiplier;
import appeng.api.networking.GridHelper;
import appeng.api.networking.IGridNode;
import appeng.api.networking.IGridNodeListener;
import appeng.api.networking.IInWorldGridNodeHost;
import appeng.api.networking.IManagedGridNode;
import appeng.api.util.AECableType;
import appeng.api.util.AEColor;
import li.cil.oc.Settings;
import li.cil.oc.common.blockentity.traits.PowerAcceptor;
import li.cil.oc.common.blockentity.traits.PowerAcceptorIntegration;
import li.cil.oc.integration.util.Power$;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.block.entity.BlockEntity;

import java.util.EnumSet;
import java.util.WeakHashMap;
import java.util.function.Consumer;

public final class AppliedEnergistics2 implements PowerAcceptorIntegration {
    public static final AppliedEnergistics2 INSTANCE = new AppliedEnergistics2();
    private final WeakHashMap<PowerAcceptor, NodeHost> hosts = new WeakHashMap<>();

    private AppliedEnergistics2() {
    }

    public IInWorldGridNodeHost getNodeHost(final PowerAcceptor host) {
        return nodeHost(host);
    }

    @Override
    public void initialize(final PowerAcceptor host) {
        final BlockEntity blockEntity = blockEntity(host);
        if (blockEntity.getLevel() != null && !blockEntity.getLevel().isClientSide) {
            nodeHost(host).scheduleCreate();
        }
    }

    @Override
    public void update(final PowerAcceptor host) {
        final BlockEntity blockEntity = blockEntity(host);
        if (blockEntity.getLevel() == null || blockEntity.getLevel().isClientSide || blockEntity.isRemoved()
                || blockEntity.getLevel().getGameTime() % Settings.get().tickFrequency() != 0) {
            return;
        }
        nodeHost(host).update();
    }

    @Override
    public void dispose(final PowerAcceptor host) {
        final NodeHost nodeHost;
        synchronized (hosts) {
            nodeHost = hosts.remove(host);
        }
        if (nodeHost != null) {
            nodeHost.destroy();
        }
    }

    @Override
    public void load(final PowerAcceptor host, final CompoundTag nbt, final net.minecraft.core.HolderLookup.Provider provider) {
        if (blockEntity(host).getLevel() != null && !blockEntity(host).getLevel().isClientSide) {
            nodeHost(host).loadFromNBT(nbt);
        }
    }

    @Override
    public void save(final PowerAcceptor host, final CompoundTag nbt, final net.minecraft.core.HolderLookup.Provider provider) {
        if (blockEntity(host).getLevel() != null && !blockEntity(host).getLevel().isClientSide) {
            nodeHost(host).saveToNBT(nbt);
        }
    }

    private NodeHost nodeHost(final PowerAcceptor host) {
        synchronized (hosts) {
            return hosts.computeIfAbsent(host, NodeHost::new);
        }
    }

    private static BlockEntity blockEntity(final PowerAcceptor host) {
        return (BlockEntity) (Object) host;
    }

    private static final class NodeHost implements IInWorldGridNodeHost, IGridNodeListener<NodeHost> {
        private final PowerAcceptor owner;
        private boolean createScheduled;
        private final IManagedGridNode managedNode;

        private NodeHost(final PowerAcceptor owner) {
            this.owner = owner;
            managedNode = GridHelper.createManagedNode(this, this)
                    .setInWorldNode(true)
                    .setGridColor(AEColor.TRANSPARENT)
                    .setIdlePowerUsage(0)
                    .setTagName(Settings.namespace() + "ae2power");
        }

        private void scheduleCreate() {
            if (!createScheduled && !managedNode.isReady()) {
                refreshExposedSides();
                createScheduled = true;
                GridHelper.onFirstTick(blockEntity(owner), value -> {
                    if (!value.isRemoved() && value.getLevel() != null && !value.getLevel().isClientSide) {
                        managedNode.create(value.getLevel(), value.getBlockPos());
                    }
                });
            }
        }

        private void update() {
            if (!managedNode.isReady()) {
                scheduleCreate();
                return;
            }

            refreshExposedSides();
            final var grid = managedNode.getGrid();
            if (grid == null) {
                return;
            }

            double budget = owner.energyThroughput() * Settings.get().tickFrequency();
            for (final Direction side : Direction.values()) {
                final double demand = Power$.MODULE$.toAE(Math.min(budget, owner.globalDemand(side)));
                if (demand > 1) {
                    final double supplied = Power$.MODULE$.fromAE(
                            grid.getEnergyService().extractAEPower(demand, Actionable.MODULATE, PowerMultiplier.CONFIG));
                    if (supplied > 0) {
                        budget -= owner.tryChangeBuffer(side, supplied, true);
                    }
                }
            }
        }

        private void loadFromNBT(final CompoundTag nbt) {
            managedNode.loadFromNBT(nbt);
        }

        private void saveToNBT(final CompoundTag nbt) {
            managedNode.saveToNBT(nbt);
        }

        private void destroy() {
            managedNode.destroy();
        }

        private void refreshExposedSides() {
            final EnumSet<Direction> sides = EnumSet.noneOf(Direction.class);
            for (final Direction side : Direction.values()) {
                if (owner.canConnectPower(side)) {
                    sides.add(side);
                }
            }
            managedNode.setExposedOnSides(sides);
        }

        @Override
        public IGridNode getGridNode(final Direction direction) {
            refreshExposedSides();
            return owner.canConnectPower(direction) ? managedNode.getNode() : null;
        }

        @Override
        public AECableType getCableConnectionType(final Direction direction) {
            return AECableType.SMART;
        }

        @Override
        public void onSaveChanges(final NodeHost nodeOwner, final IGridNode node) {
            blockEntity(owner).setChanged();
        }
    }
}
